pkg/frontend/openshiftclusterkubeconfigcredentials_post.go

```diff
@@ -0,0 +1,77 @@
+)
+
+func (f *frontend) postOpenShiftClusterKubeConfigCredentials(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	vars := mux.Vars(r)
+
+	if f.apis[vars["api-version"]].OpenShiftClusterAdminKubeconfigConverter == nil {
+		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidResourceType, "", "The resource type '%s' could not be found in the namespace '%s' for api version '%s'.", vars["resourceType"], vars["resourceProviderNamespace"], vars["api-version"])
+		return
+	}
+
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	if len(body) > 0 && !json.Valid(body) {
+		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized.")
+		return
+	}
+
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	b, err := f._postOpenShiftClusterKubeConfigCredentials(ctx, r, f.apis[vars["api-version"]].OpenShiftClusterAdminKubeconfigConverter())
+
+	reply(log, w, nil, b, err)
+}
+
+func (f *frontend) _postOpenShiftClusterKubeConfigCredentials(ctx context.Context, r *http.Request, converter api.OpenShiftClusterAdminKubeconfigConverter) ([]byte, error) {
+	vars := mux.Vars(r)
+
+	subDoc, err := f.validateSubscriptionState(ctx, r.URL.Path, api.SubscriptionStateRegistered)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(mjudeikis): Remove this once all this is communicated to the customers and this
+	// becomes defacto standard
+	if !feature.IsRegisteredForFeature(subDoc.Subscription.Properties, api.FeatureFlagAdminKubeconfig) {
+		return nil, api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "Subscription feature flag '%s' is not enabled on this subscription to use this API.", api.FeatureFlagAdminKubeconfig)
+	}
+
+	doc, err := f.dbOpenShiftClusters.Get(ctx, r.URL.Path)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return nil, api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	case err != nil:
+		return nil, err
+	}
+
+	if doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateCreating ||
+		doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateDeleting ||
+		doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateFailed && doc.OpenShiftCluster.Properties.FailedProvisioningState == api.ProvisioningStateCreating ||
+		doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateFailed && doc.OpenShiftCluster.Properties.FailedProvisioningState == api.ProvisioningStateDeleting {
+		return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeRequestNotAllowed, "", "Request is not allowed in provisioningState '%s'.", doc.OpenShiftCluster.Properties.ProvisioningState)
```

Comment:
Sorry for being late on this, but this error message would just say "Request is not allowed in provisioningState 'Failed'" if it was Failed/Creating or Failed/Deleting. This message is inaccurate, since it _would_ be allowed in Failed/Updating. We might want to include the FailedProvisioningState if it's present.

Reply:
@nilsanderselde this is a customer facing error and there is no such thing as `FailedProvisioningState` from the customer perspective: we do not expose it in our API.

> This message is inaccurate, since it would be allowed in Failed/Updating.

I however agree with this. I think [list credentials](https://github.com/Azure/ARO-RP/blob/2aba619c85ecf79459f5c9b4a3b2152ab07faf99/pkg/frontend/openshiftclustercredentials_post.go) has this problem as well. We might consider doing something similar to [put or patch](https://github.com/Azure/ARO-RP/blob/2aba619c85ecf79459f5c9b4a3b2152ab07faf99/pkg/frontend/openshiftcluster_putorpatch.go#L96-L108) where the message is more human readable.

Would you like to create a work item (or a PR straight away) to improve UX?

Reply:
TBH I think this might change signatures of the calls returns. I dont think this is worth time and effort now :/

Reply:
@mjudeikis in put or patch we handle it by returning more sepecific error messages to the customers: "Request is not allowed on cluster whose creation failed. Delete the cluster." and "Request is not allowed on cluster whose deletion failed. Delete the cluster.". I think it is just a matter changing the error message in `api.NewCloudError(...)`.

I don't think it is going to change signatures.